drv/stm32h7-fmc-demo-server/src/main.rs

```diff
@@ -0,0 +1,343 @@
+        w.faccen().clear_bit();
+
+        // Configure the memory as PSRAM (TODO: verify)
+        unsafe {
+            w.mtyp().bits(0b01);
+        }
+
+        // Turn on the memory bank.
+        w.mbken().set_bit();
+
+        // The following fields are being deliberately left in their reset
+        // states:
+        // - FMCEN is being left off
+        // - BMAP default (no remapping) is retained
+        // - Write FIFO is being left on (TODO is this correct?)
+        // - CPSIZE is being left with no special behavior on page-crossing
+        // - ASYNCWAIT is being left off since we're synchronous
+        // - EXTMOD is being left off, since it seems to only affect async
+        // - WAITCFG is being left default (TODO tweak later)
+        // - WAITPOL is treating NWAIT as active low (could change if desired)
+        // - MWID is being left at a 16 bit data bus.
+        // - MUXEN is being left with a multiplexed A/D bus.
+
+        w
+    });
+
+    // Bank timings!
+
+    // Synchronous access write/read latency, minus 2. That is, 0 means 2 cycle
+    // latency. Max value: 15 (for 17 cycles). NWAIT is not sampled until this
+    // period has elapsed, so if you're handshaking with a device using NWAIT,
+    // you almost certainly want this to be 0.
+    const DATLAT: u8 = 0;
+    // FMC_CLK division ratio relative to input (AHB3) clock, minus 1. Range:
+    // 1..=15.
+    const CLKDIV: u8 = 15; // /4, for 50 MHz
+
+    // Bus turnaround time in FMC_CLK cycles, 0..=15
+    const BUSTURN: u8 = 0;
+
+    fmc.btr1.write(|w| {
+        unsafe {
+            w.datlat().bits(DATLAT);
+        }
+        unsafe {
+            w.clkdiv().bits(CLKDIV);
+        }
+        unsafe {
+            w.busturn().bits(BUSTURN);
+        }
+
+        // Deliberately left in reset state and/or ignored:
+        // - ACCMOD: only applies when EXTMOD is set in BCR above; also probably
+        //   async only
+        // - DATAST: async only
+        // - ADDHLD: async only
+        // - ADDSET: async only
+        //
+        w
+    });
+
+    // BWTR1 register is irrelevant if we're not using EXTMOD, which we're not,
+    // currently.
+
+    // Turn on the controller.
+    fmc.bcr1.modify(|_, w| w.fmcen().set_bit());
+
+    // Fire up a server.
+    let mut server = ServerImpl { sys, fmc };
+    let mut buffer = [0; idl::INCOMING_SIZE];
+    loop {
+        idol_runtime::dispatch(&mut buffer, &mut server);
+    }
+}
+
+struct ServerImpl {
+    fmc: &'static device::fmc::RegisterBlock,
+    sys: sys_api::Sys,
+}
+
+impl NotificationHandler for ServerImpl {
+    fn current_notification_mask(&self) -> u32 {
+        0
+    }
+
+    fn handle_notification(&mut self, _bits: u32) {
+        unreachable!()
+    }
+}
+
+impl idl::InOrderFmcDemoImpl for ServerImpl {
+    fn peek16(
+        &mut self,
+        _msg: &RecvMessage,
+        addr: u32,
+    ) -> Result<u16, RequestError<Infallible>> {
+        let ptr = addr as *const u16;
```

Comment:
presumably pointer provenance would eventually prefer us to use [`ptr::from_exposed_addr`](https://doc.rust-lang.org/std/ptr/fn.from_exposed_addr.html) for ... all of these shenanigans, but that's not real yet...